File: android/app/src/main/java/com/coinswallet/coins/BitcoinCoin.java
// BitcoinCoin.java
package com.coinswallet.coins;

import com.coinswallet.CoinFactory;

import wallet.core.jni.HDWallet;
import wallet.core.jni.CoinType;
import wallet.core.jni.PrivateKey;
import wallet.core.jni.BitcoinSigHashType;
import wallet.core.jni.BitcoinScript;
import wallet.core.jni.BitcoinAddress;
import wallet.core.jni.Purpose;
import java.util.Base64;
import java.nio.charset.StandardCharsets;
import android.util.Log;

public class BitcoinCoin extends CoinFactory.Coin {
    private final HDWallet wallet;
    private int addressIndex = 0;

    public BitcoinCoin(String mnemonic) {
        super(mnemonic);
        this.wallet = super.wallet;
    }

    @Override
    public String getNewAddress() {
        // Derive a new private key for each call
        PrivateKey privateKey = wallet.getKeyForCoin(CoinType.BITCOIN);
        // Derive the address from the private key
        String address = CoinType.BITCOIN.deriveAddress(privateKey);
        addressIndex++;
        return address;
    }

    @Override
    public String getPrivateKey() {
        byte[] privateKeyBytes = wallet.getKeyForCoin(CoinType.BITCOIN).data();
        return new String(privateKeyBytes, StandardCharsets.UTF_8);
    }

    @Override
    public String signTransaction(String rawData) {
        // You need to implement this method according to the specific requirements of
        // Bitcoin transactions.
        return null;
    }
}

File: ios/BitcoinCoin.swift
//
//  BitcoinCoin.swift
//  coinswallet
//
//  Created by Tal Grynbaum on 30/06/2023.
//


import Foundation
import WalletCore

class BitcoinCoin: CoinFactory.Coin {
    private var addressIndex: Int = 0
    
    override init(mnemonic: String) {
        super.init(mnemonic: mnemonic)
    }
    
    override func getNewAddress() -> String {
        let privateKey = wallet.getKeyForCoin(coin: .bitcoin)
        let address = CoinType.bitcoin.deriveAddress(privateKey: privateKey)
        addressIndex += 1
        return address
    }
    
    override func getPrivateKey() -> String {
        let privateKeyBytes = wallet.getKeyForCoin(coin: .bitcoin).data
        return String(data: privateKeyBytes, encoding: .utf8) ?? ""
    }
    
    override func signTransaction(rawData: String) -> String {
        // Implement the logic to sign a Bitcoin transaction using the provided raw data
        return ""
    }
}
